### PR TITLE
Doc: the setup stack

### PR DESCRIPTION
**Pull request by Kenneth-T-Moore**

### CODE REVIEW

File: openmdao/docs/features/core_features/grouping_components/configure_method.rst [hunk 1]
```diff
@@ -26,19 +26,5 @@ system.
...
- - Assign Jacobians at group level
- - Set system execution order
-
-**configure**
-
- - Assign linear and nonlinear solvers to subsystems
- - Change solver settings in subsystems
- - Assign Jacobians to subsystems
- - Set execution order in subsystems
+To understand when to use setup and when to use configure, see the :ref:`Theory Manual entry on how the setup stack.<theory_setup_stack>`.
```
Comment (kmarsteller) [inline openmdao/docs/features/core_features/grouping_components/configure_method.rst]: "see the Theory Manual entry on how the setup stack." seems like a sentence fragment, assume you want to add a word like "works" to the end?

File: openmdao/docs/theory_manual/setup_stack.rst [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+**setup**
+
+ - Add subsystems
+ - Issue connections
+ - Assign linear and nonlinear solvers at this group level
+ - Change solver settings for any solver at this group level
+ - Assign Jacobians at this group level
+ - Set system execution order
+ - Add desvars, objectives, and constraints
+ - Add a case recorder to the group or a solver in that group.
```
Comment (kmarsteller) [inline openmdao/docs/theory_manual/setup_stack.rst]: or driver?

File: openmdao/docs/theory_manual/setup_stack.rst [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+**configure**
+
+ - Assign linear and nonlinear solvers to subsystems
+ - Change solver settings in subsystems
+ - Assign Jacobians to subsystems
+ - Set execution order in subsystems
+ - Add a case recorder to a subsystem or a solver in a subsystem.
+
+**Things you should never do in configure**
```
Comment (kmarsteller) [inline openmdao/docs/theory_manual/setup_stack.rst]: Perhaps mention why you should never do this? Consequences and/or effects might help someone recognize a problem they're having?

File: openmdao/docs/theory_manual/setup_stack.rst [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+setup method in each group is called recursively from top to bottom of the hierarchy. For example,
+a group may contain several components and groups. Setup is first called in that top group, during
+which, those components and groups are instantiated. However, the setup methods belonging to those sub-components
+and groups cannot be called until the top group's setup finishes. This means they are in a state where
+components and groups that are declared in the subgroup don't exist yet.
+
+To remedy this, there is a second api method called `configure` that lets you make changes to your subsystems
+after they have been created. The `configure` method is only needed with groups, and it is called
+recursively from the bottom of the hierarchy to the top, so that at any level, you can be sure that
+configure has already run for all your subsystems. This assures that changes made in higher level assemblies
```
Comment (kmarsteller) [inline openmdao/docs/theory_manual/setup_stack.rst]: Minor comment...if you are going in to make other changes, and agree with this preference, I like hyphenating multi-word adjectives like "higher-level assemblies" and "lower-level ones".  It's not incorrect to leave it as-is, though.

File: openmdao/docs/theory_manual/setup_stack.rst [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+**Things you should never do in configure**
+
+ - Add subsystems
+ - Delete subsystems
+
+Problem setup and final_setup
+-----------------------------
+
+OpenMDAO 2.0 introduces a new change to the setup process in which the original monolithic process
+is split into two separate phases triggered by the methods: `setup` and `final_setup`. The `final_setup` method is
```
Comment (kmarsteller) [inline openmdao/docs/theory_manual/setup_stack.rst]: typo: setwp == setup

File: openmdao/docs/theory_manual/setup_stack.rst [hunk 6]
```diff
@@ -0,0 +1,120 @@
+.. _theory_setup_stack:
+
+*********************************************************************
+The System Setup Stack: Understanding When to Use setup and configure
+*********************************************************************
+
+This document explains what happens during the OpenMDAO setup process, and how some of the model
```
Comment (swryan) [inline openmdao/docs/theory_manual/setup_stack.rst]: maybe "Problem setup process"?  (to be clear we are talking about the Problem setup() method, not how to install/setup OpenMDAO itself...

File: openmdao/docs/theory_manual/setup_stack.rst [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+API methods interact during that process.
+
+The purpose of the setup process is to prepare the data structures that OpenMDAO need to efficiently
```
Comment (swryan) [inline openmdao/docs/theory_manual/setup_stack.rst]: needs

File: openmdao/docs/theory_manual/setup_stack.rst [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+Problem setup and final_setup
+-----------------------------
+
+OpenMDAO 2.0 introduces a new change to the setup process in which the original monolithic process
+is split into two separate phases triggered by the methods: `setup` and `final_setup`. The `final_setup` method is
+however something you will probably never have to call, as it is called automatically the first time that
+you call `run_model` or `run_driver` after running setup. The reason that the setup process was split into two
+phases is to allow you to perform certain actions after setup:
+
+**Post setup actions**
```
Comment (kmarsteller) [inline openmdao/docs/theory_manual/setup_stack.rst]: again, I might hyphenate "Post-setup" but your tastes may vary.

File: openmdao/docs/theory_manual/setup_stack.rst [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+API methods interact during that process.
+
+The purpose of the setup process is to prepare the data structures that OpenMDAO need to efficiently
+run your model or driver. In particular, this includes setting up the vectors used for passing data
+to inputs, converging solvers, and calculating derivatives. It also includes setting up the MPI
+communicators.
+
+Setup also performs some level of model checking, mainly for critical errors. More extensive model
+checking can be done by setting "check" to True when calling `setup`, or by using the :ref:`openmdao command
+line check<om-command>`. It is recommend that you do this before your first full run of the model.
```
Comment (swryan) [inline openmdao/docs/theory_manual/setup_stack.rst]: recommended

File: openmdao/docs/theory_manual/setup_stack.rst [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+The OpenMDAO `Group` API includes three methods that are invoked during the setup process: `setup`, `configure`, and
+`initialize`. Most of the time, setup is all you need to build a group. The specific use case for
+configure is shown below in the next section. The configure method is only used for declaring options for your
```
Comment (swryan) [inline openmdao/docs/theory_manual/setup_stack.rst]: sounds like this is referring to `initialize` not `configure`... from `System`:
        self._declare_options()
        self.initialize()
        self.options.update(kwargs)

File: openmdao/docs/theory_manual/setup_stack.rst (hunk 5 above)
Comment (kmarsteller) [reply openmdao/docs/theory_manual/setup_stack.rst]: looks like this was fixed while I was doing review.

File: openmdao/docs/theory_manual/setup_stack.rst (hunk 2 above)
Comment (Kenneth-T-Moore) [reply openmdao/docs/theory_manual/setup_stack.rst]: You actually can't touch drivers from here.

File: openmdao/docs/theory_manual/setup_stack.rst [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+ - Change settings on solvers
+ - Change options on systems
+ - Add recorders
+ - Assign Jacobians
+ - Add training data to metamodels
+
+If you do anything that changes the model hierarchy, such as adding a component to a group, then you will need to
+run `setup` again.
+
+During setup, the following things happen:
```
[the remark this answers is not in the document]
Comment (Kenneth-T-Moore) [reply openmdao/docs/theory_manual/setup_stack.rst]: All issues have been addresed

**Pull request merged**